File: file_reader.py
"""
File Reader Utilities

This module contains utilities for reading files and performing linear fits.
Ported from Fortran to Python.

Author: Victor Carreira (Original Fortran)
Python Port: 2026
"""

import numpy as np
from typing import Tuple, List


def skip_line(file_handle) -> None:
    """
    Skip the next line in a file.
    
    Parameters
    ----------
    file_handle : file object
        Open file handle
    """
    file_handle.readline()


def count_lines(filename: str, skip_header: bool = True) -> int:
    """
    Count the number of lines in a file.
    
    Parameters
    ----------
    filename : str
        Path to the file
    skip_header : bool, optional
        Whether to skip the first line (default: True)
    
    Returns
    -------
    int
        Number of lines in the file
    """
    with open(filename, 'r') as f:
        if skip_header:
            f.readline()  # Skip header
        
        nlines = 0
        for line in f:
            nlines += 1
    
    return nlines


def read_line(file_handle) -> str:
    """
    Read the next line from a file.
    
    Parameters
    ----------
    file_handle : file object
        Open file handle
    
    Returns
    -------
    str
        The line read from the file
    """
    return file_handle.readline().strip()


def read_data_file(filename: str, skip_header: bool = True) -> Tuple[np.ndarray, np.ndarray]:
    """
    Read x and y data from a file.
    
    Parameters
    ----------
    filename : str
        Path to the data file
    skip_header : bool, optional
        Whether to skip the first line (default: True)
    
    Returns
    -------
    x, y : np.ndarray
        Arrays containing the x and y data
    """
    data = []
    
    with open(filename, 'r') as f:
        if skip_header:
            f.readline()  # Skip header
        
        for line in f:
            parts = line.strip().split()
            if len(parts) >= 2:
                x_val = float(parts[0])
                y_val = float(parts[1])
                data.append([x_val, y_val])
    
    data_array = np.array(data)
    return data_array[:, 0], data_array[:, 1]


def fit_ab(x: np.ndarray, y: np.ndarray) -> Tuple[float, float, float, float]:
    """
    Perform linear fit to calculate coefficients a and b for y = a*x + b.
    Also calculate uncertainties sa and sb.
    
    Parameters
    ----------
    x : np.ndarray
        X data values
    y : np.ndarray
        Y data values
    
    Returns
    -------
    a, b, sa, sb : float
        Linear (a) and angular (b) coefficients and their uncertainties
    """
    n = len(x)
    
    if n != len(y):
        raise ValueError('x and y arrays must have the same length')
    
    if n < 2:
        raise ValueError('Need at least 2 data points for fitting')
    
    # Calculate delta, mi, and sigma
    sum_x = np.sum(x)
    sum_y = np.sum(y)
    sum_xx = np.sum(x * x)
    sum_xy = np.sum(x * y)
    
    delta = n * sum_xx - sum_x**2
    
    if delta == 0:
        raise ValueError('Delta is zero - cannot perform fit')
    
    mi = sum_x / n
    sig2 = np.sum((x - mi)**2) / n
    
    # Calculate uncertainties
    sa = np.sqrt(sig2 * n / delta)
    sb = np.sqrt(sig2 * sum_xx / delta)
    
    # Calculate coefficients using least squares
    a = (n * sum_xy - sum_x * sum_y) / delta
    b = (sum_xx * sum_y - sum_xy * sum_x) / delta
    
    return a, b, sa, sb


def process_data_file(input_file: str, output_file: str = 'params.txt',
                      skip_header: bool = True) -> Tuple[float, float, float, float]:
    """
    Read data from a file, perform linear fit, and write results to output file.
    
    Parameters
    ----------
    input_file : str
        Path to input data file
    output_file : str, optional
        Path to output file for parameters (default: 'params.txt')
    skip_header : bool, optional
        Whether to skip the first line of input file (default: True)
    
    Returns
    -------
    a, b, sa, sb : float
        Linear fit coefficients and uncertainties
    """
    # Read data
    x, y = read_data_file(input_file, skip_header=skip_header)
    
    print(f'Number of data points: {len(x)}')
    
    # Perform fit
    a, b, sa, sb = fit_ab(x, y)
    
    print(f'Coefficients and uncertainties:')
    print(f'a = {a:.6f}, sa = {sa:.6f}')
    print(f'b = {b:.6f}, sb = {sb:.6f}')
    
    # Write results to output file
    with open(output_file, 'w') as f:
        f.write('       a       b      sa     sb\n')
        f.write(f'{a:10.6f} {b:10.6f} {sa:10.6f} {sb:10.6f}\n')
    
    return a, b, sa, sb


if __name__ == '__main__':
    # Example usage
    # Assumes a file 'dados.txt' exists with header and x, y columns
    import sys
    
    if len(sys.argv) > 1:
        input_file = sys.argv[1]
    else:
        input_file = 'dados.txt'
    
    try:
        a, b, sa, sb = process_data_file(input_file)
        print(f'\nResults written to params.txt')
    except FileNotFoundError:
        print(f'Error: File {input_file} not found')
    except Exception as e:
        print(f'Error: {e}')

File: src/file_reader.rs
//! File Reader Utilities
//!
//! This module contains utilities for reading files and performing linear fits.
//! Ported from Fortran to Rust.
//!
//! Author: Victor Carreira (Original Fortran)
//! Rust Port: 2026

use std::fs::File;
use std::io::{BufRead, BufReader, Write};

/// Count the number of lines in a file.
///
/// # Arguments
///
/// * `filename` - Path to the file
/// * `skip_header` - Whether to skip the first line
///
/// # Returns
///
/// Number of lines in the file
pub fn count_lines(filename: &str, skip_header: bool) -> Result<usize, std::io::Error> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    let mut lines = reader.lines();

    if skip_header {
        lines.next();
    }

    let count = lines.count();
    Ok(count)
}

/// Read x and y data from a file.
///
/// # Arguments
///
/// * `filename` - Path to the data file
/// * `skip_header` - Whether to skip the first line
///
/// # Returns
///
/// Tuple of (x_values, y_values) as vectors
pub fn read_data_file(
    filename: &str,
    skip_header: bool,
) -> Result<(Vec<f64>, Vec<f64>), Box<dyn std::error::Error>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    let mut lines = reader.lines();

    if skip_header {
        lines.next();
    }

    let mut x_values = Vec::new();
    let mut y_values = Vec::new();

    for line_result in lines {
        let line = line_result?;
        let parts: Vec<&str> = line.split_whitespace().collect();

        if parts.len() >= 2 {
            let x: f64 = parts[0].parse()?;
            let y: f64 = parts[1].parse()?;
            x_values.push(x);
            y_values.push(y);
        }
    }

    Ok((x_values, y_values))
}

/// Perform linear fit to calculate coefficients a and b for y = a*x + b.
/// Also calculate uncertainties sa and sb.
///
/// # Arguments
///
/// * `x` - X data values
/// * `y` - Y data values
///
/// # Returns
///
/// Tuple of (a, b, sa, sb) - Linear (a) and angular (b) coefficients and their uncertainties
pub fn fit_ab(x: &[f64], y: &[f64]) -> Result<(f64, f64, f64, f64), &'static str> {
    let n = x.len();

    if n != y.len() {
        return Err("x and y arrays must have the same length");
    }

    if n < 2 {
        return Err("Need at least 2 data points for fitting");
    }

    let n_f64 = n as f64;

    // Calculate sums
    let sum_x: f64 = x.iter().sum();
    let sum_y: f64 = y.iter().sum();
    let sum_xx: f64 = x.iter().map(|&xi| xi * xi).sum();
    let sum_xy: f64 = x.iter().zip(y.iter()).map(|(&xi, &yi)| xi * yi).sum();

    let delta = n_f64 * sum_xx - sum_x * sum_x;

    if delta == 0.0 {
        return Err("Delta is zero - cannot perform fit");
    }

    let mi = sum_x / n_f64;
    let sig2: f64 = x.iter().map(|&xi| (xi - mi).powi(2)).sum::<f64>() / n_f64;

    // Calculate uncertainties
    let sa = (sig2 * n_f64 / delta).sqrt();
    let sb = (sig2 * sum_xx / delta).sqrt();

    // Calculate coefficients using least squares
    let a = (n_f64 * sum_xy - sum_x * sum_y) / delta;
    let b = (sum_xx * sum_y - sum_xy * sum_x) / delta;

    Ok((a, b, sa, sb))
}

/// Read data from a file, perform linear fit, and write results to output file.
///
/// # Arguments
///
/// * `input_file` - Path to input data file
/// * `output_file` - Path to output file for parameters
/// * `skip_header` - Whether to skip the first line of input file
///
/// # Returns
///
/// Tuple of (a, b, sa, sb) - Linear fit coefficients and uncertainties
pub fn process_data_file(
    input_file: &str,
    output_file: &str,
    skip_header: bool,
) -> Result<(f64, f64, f64, f64), Box<dyn std::error::Error>> {
    // Read data
    let (x, y) = read_data_file(input_file, skip_header)?;

    println!("Number of data points: {}", x.len());

    // Perform fit
    let (a, b, sa, sb) = fit_ab(&x, &y)?;

    println!("Coefficients and uncertainties:");
    println!("a = {:.6}, sa = {:.6}", a, sa);
    println!("b = {:.6}, sb = {:.6}", b, sb);

    // Write results to output file
    let mut file = File::create(output_file)?;
    writeln!(file, "       a       b      sa     sb")?;
    writeln!(file, "{:10.6} {:10.6} {:10.6} {:10.6}", a, b, sa, sb)?;

    Ok((a, b, sa, sb))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fit_ab() {
        // Test with simple linear data: y = 2x + 1
        let x = vec![0.0, 1.0, 2.0, 3.0, 4.0];
        let y = vec![1.0, 3.0, 5.0, 7.0, 9.0];

        let result = fit_ab(&x, &y);
        assert!(result.is_ok());

        let (a, b, _sa, _sb) = result.unwrap();
        // a should be approximately 2.0
        assert!((a - 2.0).abs() < 1e-10);
        // b should be approximately 1.0
        assert!((b - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_fit_ab_error_cases() {
        let x = vec![1.0];
        let y = vec![1.0];

        // Should fail with too few points
        assert!(fit_ab(&x, &y).is_err());

        let x = vec![1.0, 2.0];
        let y = vec![1.0];

        // Should fail with mismatched lengths
        assert!(fit_ab(&x, &y).is_err());
    }
}
